crates/core/tedge_mapper/src/c8y/tests.rs

```diff
@@ -1124,6 +1124,149 @@ async fn mapper_handles_multiline_sm_requests() {
+    )
+    .await;
+    sm_mapper.abort();
+}
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+#[serial]
+async fn mapper_publishes_child_device_create_message() {
+    // The test assures tedge-mapper checks if there is a directory for operations for child devices, then it reads and
+    // correctly publishes the child device create message on to `c8y/s/us`
+    // and verifies the device create message.
+    let broker = mqtt_tests::test_mqtt_broker();
+    let mut messages = broker.messages_published_on("c8y/s/us").await;
+
+    let (_tmp_dir, _ops_dir, sm_mapper) = start_c8y_mapper(broker.port).await.unwrap();
+
+    publish_a_fake_jwt_token(broker).await;
+
+    // Expect smartrest message on `c8y/s/us` with expected payload "101,child1,child1,thin-edge.io-child".
+    mqtt_tests::assert_received_all_expected(
+        &mut messages,
+        TEST_TIMEOUT_MS,
+        &["101,child1,child1,thin-edge.io-child"],
+    )
+    .await;
+    sm_mapper.abort();
+}
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+#[serial]
+async fn mapper_publishes_supported_operations_for_child_device() {
+    // The test assures tedge-mapper checks if there is a directory for operations for child devices, then it reads and
+    // correctly publishes the child device create message on to `c8y/s/us/child1`
+    // and verifies the device create message.
+    let broker = mqtt_tests::test_mqtt_broker();
+    let mut messages = broker.messages_published_on("c8y/s/us/child1").await;
+
+    let (_tmp_dir, _ops_dir, sm_mapper) = start_c8y_mapper(broker.port).await.unwrap();
+
+    publish_a_fake_jwt_token(broker).await;
+
+    // Expect smartrest message on `c8y/s/us/child1` with expected payload "114,c8y_ChildTestOp1,c8y_ChildTestOp2.
+    mqtt_tests::assert_received_all_expected(
+        &mut messages,
+        TEST_TIMEOUT_MS,
+        &["114,c8y_ChildTestOp1,c8y_ChildTestOp2\n"],
+    )
+    .await;
+    sm_mapper.abort();
+}
+
+#[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+#[serial]
+async fn mapper_publishes_dynamically_supported_operations_for_tedge_device() {
+    // The test assures tedge-mapper checks if there are operations, then it reads and
+    // correctly publishes them on to `c8y/s/us`.
+    // When mapper is running test adds a new operation into the operations directory, then the mapper discovers the new
+    // operation and publishes list of supported operation including the new operation, and verifies the device create message.
+    let broker = mqtt_tests::test_mqtt_broker();
+    let mut messages = broker.messages_published_on("c8y/s/us").await;
+
+    let (_tmp_dir, ops_dir, sm_mapper) = start_c8y_mapper(broker.port).await.unwrap();
+
+    publish_a_fake_jwt_token(broker).await;
+    // Wait for the mapper to start properly and start the wacher for the directories
+
+    // Add an operation dynamically
+    ops_dir.dir("operations").dir("c8y").file("c8y_TestOp3");
```

Comment:
This wait logic can be made more deterministic by waiting for the initial `114` message sent by the mapper for the parent device. That way, you don't have to resort to this looped file creation to catchup. That initial `114` message guarantees that the mapper has started watching the directory. You can spawn this task after that wait.

Reply:
Using the `health-check` message to make sure the `tedge-mapper` as started properly and added the `watcher` to the directories.